docs/meta.rst

```diff
@@ -17,55 +17,124 @@ Taichi kernels are *lazily instantiated* and a lot of computation can happen at
 ------------------------------------------------------------
 
+However, the ``copy`` template shown above is not perfect, i.e., it can only be
+used to copy 1D tensors. What if we want to copy 2D tensors? Do we have to write
+another kernel?
+
+.. code-block:: python
+
+    @ti.kernel
+    def copy2d(x: ti.template(), y: ti.template()):
+        for i, j in x:
+            y[i, j] = x[i, j]
+
+Not necessary! Taichi provides ``ti.grouped`` syntax which enable you to get
+loop indices into a grouped vector, therefore unify different dimensionalities.
+For example:
+
 .. code-block:: python
 
     @ti.kernel
     def copy(x: ti.template(), y: ti.template()):
         for I in ti.grouped(y):
+            # I is a vector with same dimensionality with x and data type i32
+            # If y is 0D, then I = None
+            # If y is 1D, then I = ti.Vector([i])
+            # If y is 2D, then I = ti.Vector([i, j])
+            # If y is 3D, then I = ti.Vector([i, j, k])
+            # ...
             x[I] = y[I]
 
     @ti.kernel
     def array_op(x: ti.template(), y: ti.template()):
-        # If tensor x is 2D
-        for I in ti.grouped(x): # I is a vector of size x.dim() and data type i32
+        # if tensor x is 2D:
+        for I in ti.grouped(x): # I is simply a 2D vector with x data type i32
             y[I + ti.Vector([0, 1])] = I[0] + I[1]
-        # is equivalent to
+
+        # then it is equivalent to:
         for i, j in x:
             y[i, j + 1] = i + j
 
-Tensor size reflection
-----------------------
 
-Sometimes it will be useful to get the dimensionality (``tensor.dim()``) and shape (``tensor.shape()``) of tensors.
-These functions can be used in both Taichi kernels and python scripts.
+Tensor meta data
+----------------
+
+Sometimes it will be useful to get the data type (``tensor.dtype``) and shape (``tensor.shape``) of tensors.
+These attributes can be accessed in both Taichi kernels and python scripts.
 
 .. code-block:: python
 
   @ti.func
-  def print_tensor_size(x: ti.template()):
-    print(x.dim())
-    for i in ti.static(range(x.dim())):
-      print(x.shape()[i])
+  def print_tensor_info(x: ti.template()):
+    print('Tensor dimensionality is', len(x.shape))
+    for i in ti.static(range(len(x.shape))):
+      print('Size alone dimension', i, 'is', x.shape[i])
+    ti.static_print('Tensor data type is', x.dtype)
+
+See :ref:`scalar_tensor` for more details.
+
+.. note::
+
+    For sparse tensors, the full domain shape will be returned.
+
+
+Matrix & vector meta data
+-------------------------
+
+Sometimes it will also be useful to get the matrix column and row numbers when
+you want to write dimensionality-independent code, such as reusing code between
+2D/3D physical simulations.
+
+``matrix.m`` equals to the column number of matrix, while ``matrix.n`` equals to
+the row number of matrix.
+Since vectors are considered as matrices with one column, ``vector.n`` is simply
+the dimensionality of vector.
```

Comment:
```suggestion
Getting the number of matrix columns and rows will allow
you to write dimensionality-independent code. For example, this can be used to unify
2D and 3D physical simulators.

``matrix.m`` equals to the number of columns of a matrix, while ``matrix.n`` equals to
the number of rows of a matrix.
Since vectors are considered as matrices with one column, ``vector.n`` is simply
the dimensionality of the vector.
```